lib/rtrace/umd/index.js

```diff
@@ -0,0 +1,384 @@
+// GENERATED FILE. DO NOT EDIT.
+var rtrace = (function(exports) {
+  "use strict";
+  
+  Object.defineProperty(exports, "__esModule", {
+    value: true
+  });
+  exports.default = exports.Rtrace = void 0;
+  // WebAssembly pages are 65536 kb
+  const PAGE_SIZE_BITS = 16;
+  const PAGE_SIZE = 1 << PAGE_SIZE_BITS;
+  const PAGE_MASK = PAGE_SIZE - 1; // Wasm32 pointer size is 4 bytes
+  
+  const PTR_SIZE_BITS = 2;
+  const PTR_SIZE = 1 << PTR_SIZE_BITS;
+  const PTR_MASK = PTR_SIZE - 1;
+  const PTR_VIEW = Uint32Array;
+  const BLOCK_OVERHEAD = PTR_SIZE;
+  
+  function assert(x) {
+    if (!x) throw Error("assertion failed");
+    return x;
+  }
+  
+  Error.stackTraceLimit = 50;
+  
+  function trimStacktrace(stack, levels) {
+    return stack.split(/\r?\n/).slice(1 + levels);
+  }
+  
+  function isTLSF(stack) {
+    return stack[0].startsWith("    at ~lib/rt/tlsf/");
+  }
+  
+  class Rtrace {
+    constructor(options) {
+      this.options = options || {};
+  
+      this.onerror = this.options.onerror || function () {
+        /* nop */
+      };
+  
+      this.oninfo = this.options.oninfo || function () {
+        /* nop */
+      };
+  
+      this.memory = null;
+      this.shadow = null;
+      this.shadowStart = 0x100000000;
+      this.refCounts = new Map();
+      this.blocks = new Map();
+      this.allocSites = new Map();
+      this.allocCount = 0;
+      this.resizeCount = 0;
+      this.moveCount = 0;
+      this.freeCount = 0;
+      this.incrementCount = 0;
+      this.decrementCount = 0; // The following hooks cannot just be on the prototype but must be
+      // bound so the Rtrace instance can be used as a WebAssembly import.
+  
+      this.onalloc = this.onalloc.bind(this);
```

Comment:
How about use arrow functions instead bindings? 
instead `this.load_val_i32 = this.load_val_i32.bind(this)` just use:
```js
load_val_i32 = (id, value) => {
   return value;
}
```
and for rest methods use the same.

Reply:
Do you mean moving all of the declarations into the ctor like that?

Reply:
No, just remove `this.load_val_i32 = this.load_val_i32.bind(this)` and friends and bind this via arrow function during declaration like:

```js
class RTrace {

   ....
   load_val_i32 = (id, value) => {
     return value;
  }
  ....
}
```
instead:
```js
class RTrace {
   constructor() {
       this.load_val_i32 = this.load_val_i32.bind(this);
   }
   ....
   load_val_i32(id, value) {
     return value;
  }
}
```

Reply:
Hmm, how would that look for

```ts
  load_ptr(id, bytes, offset, address) {
    this.accessShadow(address + offset, bytes, true);
    return address;
  }
```

when `this` is accessed?

Reply:
Also note that `bind`ing the functions makes them usable without a proper `this` context, i.e. as a Wasm import that naturally loses the `this` context. Wondering if that'd still work when not `bind`ing.

Reply:
instance methods defined via arrow function grabbed current context (this of class or object). It's typical alternative for `this.method = this.method.bind(this)`

Reply:
Hmm

```
SyntaxError: unknown: Support for the experimental syntax 'classProperties' isn't currently enabled (179:11):

  177 |   // Runtime instrumentation
  178 |
> 179 |   onalloc = (ptr) => {
```

Reply:
Hmm, class properties should work since `V8 7.2`

Reply:
Yeah, I think that's a Babel-ish error. May I suggest to make this a follow-up? Otherwise sounds good if it works!

Reply:
For babel you should just add this: https://babeljs.io/docs/en/babel-plugin-syntax-class-properties. It add only syntax for parser, but don't do actually transformation (lowering)